packages/ui-components/src/components/json-viewer/json-viewer-hook.ts

```diff
@@ -0,0 +1,141 @@
+  );
+};
+
+type UseJsonViewerProps = {
+  json: any;
+  title: string;
+  readonly?: boolean;
+  onChange?: (json: any) => void;
+  renderFileButton: (props: {
+    fileUrl: string;
+    isProductionFile: boolean;
+    handleDownloadFile: (fileUrl: string, ext?: string) => void;
+  }) => React.ReactNode;
+};
+
+export const useJsonViewer = ({
+  json,
+  title,
+  readonly = true,
+  onChange,
+  renderFileButton,
+}: UseJsonViewerProps) => {
+  const [isEditMode, setIsEditMode] = useState(!json && !readonly);
+
+  const showCopySuccessToast = () =>
+    toast({
+      title: t('Copied to clipboard'),
+      duration: COPY_PASTE_TOAST_DURATION,
+    });
+
+  const showCopyFailureToast = () =>
+    toast({
+      title: t('Failed to copy to clipboard'),
+      duration: COPY_PASTE_TOAST_DURATION,
+    });
+
+  const handleCopy = () => {
+    const text = JSON.stringify(json, null, 2);
+    if (navigator.clipboard) {
+      navigator.clipboard
+        .writeText(text)
+        .then(showCopySuccessToast)
+        .catch(showCopyFailureToast);
+    } else {
+      clipboardUtils.copyInInsecureContext({
+        text,
+        onSuccess: showCopySuccessToast,
+        onError: showCopyFailureToast,
+      });
+    }
+  };
+
+  const handleDownload = () => {
+    const blob = new Blob([JSON.stringify(json, null, 2)], {
+      type: 'application/json',
+    });
+    const url = URL.createObjectURL(blob);
+    handleDownloadFile(url);
+  };
+
+  const handleDownloadFile = (fileUrl: string, ext = '') => {
+    const link = document.createElement('a');
+    link.href = fileUrl;
+    link.download = `${title}${ext}`;
+    link.click();
+    URL.revokeObjectURL(fileUrl);
+  };
+
+  const apply = (json: string) => {
+    if (onChange) {
+      try {
+        const jsonData = JSON.parse(json);
+        onChange(jsonData);
+      } catch {
+        toast(UNSAVED_CHANGES_TOAST);
+      }
+    }
+    setIsEditMode(false);
+  };
+
+  useLayoutEffect(() => {
+    if (typeof json === 'object') {
+      const stringValuesHTML = Array.from(
+        document.getElementsByClassName('string-value'),
+      );
+
+      const stepFileUrlsHTML = stringValuesHTML.filter(
+        (el) =>
+          isStepFileUrl(el.innerHTML) ||
+          isStepFileUrl(el.parentElement!.nextElementSibling?.innerHTML),
+      );
+
+      stepFileUrlsHTML.forEach((el: Element) => {
+        const fileUrl = removeDoubleQuotes(el.innerHTML)
+          .trim()
+          .replace('\n', '');
+        el.className += ' hidden';
+
+        const rootElem = document.createElement('div');
+        const root = createRoot(rootElem);
+
+        el.parentElement!.replaceChildren(el as Node, rootElem as Node);
+        const isProductionFile = fileUrl.startsWith('file://');
+
+        root.render(
+          renderFileButton({
+            fileUrl,
+            isProductionFile,
+            handleDownloadFile,
+          }),
+        );
```

Comment:
@rSnapkoOpenOps please check that rendering the file button still works as before

Reply:
would also be good to capture it as a story